valhalla/thor/bidirectional_astar.h

```diff
@@ -152,50 +154,294 @@ class BidirectionalAStar : public PathAlgorithm {
+        // if this is a downward transition (ups are always allowed) AND we are no longer allowed OR
+        // we cant get the tile at that level (local extracts could have this problem) THEN bail
+        graph_tile_ptr trans_tile = nullptr;
+        if ((!trans->up() && hierarchy_limits[trans->endnode().level()].StopExpanding()) ||
+            !(trans_tile = graphreader.GetGraphTile(trans->endnode()))) {
+          continue;
+        }
+
+        hierarchy_limits[node.level()].up_transition_count += trans->up();
+        // setup for expansion at this level
+        const auto* trans_node = trans_tile->node(trans->endnode());
+        EdgeMetadata trans_meta =
+            EdgeMetadata::make(trans->endnode(), trans_node, trans_tile, edgestatus);
+        uint32_t trans_shortcuts = 0;
+        // expand the edges from this node at this level
+        for (uint32_t i = 0; i < trans_node->edge_count(); ++i, ++trans_meta) {
+          disable_uturn =
+              ExpandInner<expansion_direction>(graphreader, pred, opp_pred_edge, trans_node, pred_idx,
+                                               trans_meta, trans_shortcuts, trans_tile,
+                                               offset_time) ||
+              disable_uturn;
+        }
+      }
+    }
+
+    // Now, after having looked at all the edges, including edges on other levels,
+    // we can say if this is a deadend or not, and if so, evaluate the uturn-edge (if it exists)
+    if (!disable_uturn && uturn_meta) {
+      // If we found no suitable edge to add, it means we're at a deadend
+      // so lets go back and re-evaluate a potential u-turn
+      pred.set_deadend(true);
+
+      // TODO Is there a shortcut that supersedes our u-turn?
+      // Decide if we should expand a shortcut or the non-shortcut edge...
+
+      // Expand the uturn possiblity
+      disable_uturn =
+          ExpandInner<expansion_direction>(graphreader, pred, opp_pred_edge, nodeinfo, pred_idx,
+                                           uturn_meta, shortcuts, tile, offset_time) ||
+          disable_uturn;
+    }
+
+    return disable_uturn;
+  }
+
+  // Runs in the inner loop of `Expand`, essentially evaluating if
+  // the edge described in `meta` should be placed on the stack
+  // as well as doing just that.
+  //
+  // Returns false if uturns are allowed.
+  // Returns true if we will expand or have expanded from this edge. In that case we disallow uturns.
+  // Some edges we won't expand from, but we will still put them on the adjacency list in order to
+  // connect the forward and reverse paths. In that case we return false to allow uturns only if this
+  // edge is a not-thru edge that will be pruned.
+  //
+  template <const ExpansionType expansion_direction,
+            const bool FORWARD = expansion_direction == ExpansionType::forward,
+            const bool REVERSE = expansion_direction == ExpansionType::reverse>
```

Comment:
Curious as to why evaluating `REVERSE` is necessary? Most places in the function are using `!FORWARD` to mean `REVERSE`, isn't it?

Reply:
There was only one place where `REVERSE` was needed - I just replaced it with `!FORWARD` as suggested.  I'm generally not a huge fan of the `!` operator - I find it hard to spot sometimes.